server/jetstream_api.go

```diff
@@ -620,7 +625,79 @@ var allJsExports = []string{
 	JSApiConsumerDelete,
 }
 
+func (js *jetStream) apiDispatch(sub *subscription, c *client, subject, reply string, rmsg []byte) {
+	js.mu.RLock()
+	s := js.srv
+	rr := js.apiSubs.Match(subject)
+	js.mu.RUnlock()
+
+	// Shortcircuit.
+	if len(rr.psubs)+len(rr.qsubs) == 0 {
+		return
+	}
+
+	// We should only have psubs and only 1 per result.
+	// FIXME(dlc) - Should we respond here with NoResponders or error?
+	if len(rr.psubs) != 1 {
+		s.Warnf("Malformed JetStream API Request: [%s] %q", subject, rmsg)
+		return
+	}
+	jsub := rr.psubs[0]
+
+	// If this is directly from a client connection ok to do in place.
+	if c.kind == CLIENT {
+		jsub.icb(sub, c, subject, reply, rmsg)
+		return
+	}
+
+	// If we are here we have received this request over a non client connection.
+	// We need to make sure not to block. We will spin a Go routine per but also make
+	// sure we do not have too many outstanding.
+	if apiOut := atomic.AddInt64(&js.apiCalls, 1); apiOut > 1024 {
+		atomic.AddInt64(&js.apiCalls, -1)
+		ci, acc, _, msg, err := s.getRequestInfo(c, rmsg)
+		if err == nil {
+			resp := &ApiResponse{Type: JSApiOverloadedType, Error: jsInsufficientErr}
+			s.sendAPIErrResponse(ci, acc, subject, reply, string(msg), s.jsonResponse(&resp))
+		} else {
+			s.Warnf(badAPIRequestT, rmsg)
+		}
+		s.Warnf("JetStream API Overloaded: %d calls outstanding", apiOut)
+		return
+	}
+
+	// If we are here we can properly dispatch this API call.
+	// Copy the message and the client. Client for the pubArgs.
+	// FIXME(dlc) - Should cleanup eventually and make sending
+	// and receiving internal messages more formal.
+	rmsg = append(rmsg[:0:0], rmsg...)
+	client := &client{srv: s, kind: JETSTREAM}
+	client.pa = c.pa
```

Comment:
Are you sure about that? I mean the content will still reference buffer that is going to be overwritten.
`client.pa = c.pa` causes a copy since `pa` is `pubArg` which is a struct, but all `[]byte` inside (may) reference the read buffer. I don't think that this ensuring that those are copied...

Reply:
JSApi strictly only uses it to piece apart hdr and msg body. In clustered mode can not use or depend on passed in client for any more than that.

Reply:
Ok, a simple comment would help clear the confusion.

Reply:
Will do.